src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioDocumentTrackerSubscriber.cs

```diff
@@ -0,0 +1,304 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    [System.Composition.Shared]
+    [Export(typeof(VisualStudioDocumentTrackerSubscriber))]
+    internal class DefaultVisualStudioDocumentTrackerSubscriber : VisualStudioDocumentTrackerSubscriber
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly IBufferGraphFactoryService _bufferGraphService;
+        private readonly WorkspaceStateFactory _workspaceStateFactory;
+
+        [ImportingConstructor]
+        public DefaultVisualStudioDocumentTrackerSubscriber(
+            ForegroundDispatcher foregroundDispatcher,
+            IBufferGraphFactoryService bufferGraphService,
+            WorkspaceStateFactory workspaceStateFactory)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (bufferGraphService == null)
+            {
+                throw new ArgumentNullException(nameof(bufferGraphService));
+            }
+
+            if (workspaceStateFactory == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceStateFactory));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _bufferGraphService = bufferGraphService;
+            _workspaceStateFactory = workspaceStateFactory;
+        }
+
+        public override void Unsubscribe(DefaultVisualStudioDocumentTracker documentTracker)
+        {
+            if (documentTracker == null)
+            {
+                throw new ArgumentNullException(nameof(documentTracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var diskBuffer = documentTracker.TextBuffer;
+            if (diskBuffer.Properties.TryGetProperty<CSharpBufferChangeListener>(typeof(CSharpBufferChangeListener), out var csharpBufferChangeListener))
+            {
+                diskBuffer.Properties.RemoveProperty(typeof(CSharpBufferChangeListener));
+                csharpBufferChangeListener.Dispose();
+            }
+
+            documentTracker.Unsubscribe();
+        }
+
+        public override void Subscribe(DefaultVisualStudioDocumentTracker documentTracker)
+        {
+            if (documentTracker == null)
+            {
+                throw new ArgumentNullException(nameof(documentTracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (documentTracker.TextViews.Count == 0)
```

Comment:
I don't like this restriction but getting the buffer graph of just the disk buffer isn't sufficient (it doesn't encompass the C# buffer). If you have a better idea i'm all ears.